Create whitelabel builds for bluetooth enabled version of the app.

johnschoeman opened the pull request:
This pr is a handful of commits that work towards introducing a white labeling strategy to build the application for either location based or Bluetooth based contact tracing strategies.  Given that one app cannot have both bluetooth and location based data for privacy concerns we need to be able to build the app binary with different permission settings in info.plist, in order to do this we introduce two build targets in xcode, GPS and BTE, that have different info.plist and bundle ids, but share all of the other native libraries and code. We exposes the tracing strategy that the given build is using to the js side over the bridge via a 'getTracingStrategy' function.

Ive left the commits separate for ease of review and i would recommend reviewing one at a time.

Note that this work only sets up the distinction between the gps and bte based builds. It does not implement any bluetooth permission work, it will just give us some space to do so.


### Running GPS

```
yarn run-ios-gps
```

![gps](https://user-images.githubusercontent.com/16049495/82203605-8b3f7500-98d1-11ea-8c57-619e1e26affe.gif)


### Running BTE

```
yarn run-ios-bte
```

![bte](https://user-images.githubusercontent.com/16049495/82203652-94c8dd00-98d1-11ea-8231-12ab4ff8980d.gif)


## Commit 1:

Use modules imports for assets …
41fec15
Why:
Currently we are importing png and svg assets individually from
shared/assets. This leads to a fair amount of churn in the import
statements and is somewhat of a pain to maintain overtime. By bundling
the assets in a common module we can reduce a bit of cognitive overhead
by allowing designers and developers to always be importing assets from
same place, leading to some significant speed gains during iterations on
design implementations. Additionally,this will also allow us a place to
put some type safety if we ever introduce typescript to the project.
There is no technical gains to be realized with this strategy, it is
purely to make life a bit easier for designers.

This commit:
Introduces an /shared/assets/index.js to import icons and images from.
No new functionality has been added.

----

This commit is perhaps not necessary, but while getting familiar with the app I ended up using this clean up as a way to build some context, I was also concerned about missing some path updates when moving the assets folder out of shared in the following commit and this refactor made it less likely for me to make a mistake.

## Commit 2

Introduce gps and bte ios targets …
67ade64
Why:
We would like to introduce bluetooth to the application for bluetooth
based contact tracing. Given the privacy concerns of having both
location and bluetooth data available to the same application, google
and apple do not allow for both location and bluetooth to be available
at the same time.

Hence in order to service both location and bluetooth
we will need to provide separate apps with separate info.plists and
bundle identifiers. At the same time we would like to share a much
application logic between the two applications as to reduce the
maintenance and organizational overhead. Ideally as close to 100% of the
javascript layer is common and for the ios layer we only need to change
the info.plists.

This commit:

Introduces two build targets, GPS and BTE, to the xcode project which
are setup to read from the corresponding groups for config and info.plist
A COVIDSafePathsConfig file was introduced to expose a
'getTracingStrategy' function to the react native side so that the js
runtime can determine which tracing strategy the bundle will be using.
This will allow us flexibility to programmatically determine which
tracing strategy is being used either within a given react component or
to conditional render components altogether. This also allows use to
maximize the amount of shared native code.

To build the corresponding targets via the react native cli, you can
pass the --scheme flag with the corresponding target:

```
react-native run-ios --scheme "GPS"
react-native run-ios --scheme "BTE"
```

We updated the bundle ids to `org.pathcheck.covid-safepaths.gps` and
`org.pathcheck.covid-safepaths.bte` respectively. Depending on app store
considerations, it might be easier to revert
`org.pathcheck.covid-safepaths.gps` to `org.pathcheck.covid-safepaths`

Note that theres a fair amount of location information storing logic
currently in AppDelegate, a future commit might want to refactor this so
that the gps specific code is only built in the GPS target builds.

## Commit 3

ollback js code spliting …
d61cf20
Why:
Given that we are using the two build targets of GPS and BTE on the
native side, it is no longer necessary to split the code on the
javascript side.

This commit:
Moves all of the application logic from gps/app back to app/
Additionally assets where moved out of the shared folder into app/ so
that we could remove the shared/ directory altogether. This commit is a
refactor and introduces no new functionality.

## Commit 4

Refactor onboarding5 to use functional component …
4b982d0
Why:
We would like to introduce some information to the component via a
context. It is best practice to use hooks and so we would like to use
the useContext hook for this. Currently the onboarding5 component is
build with the class syntax keeping us from doing this.

This commit:
Refactors the Onbarding5 component to use the functional style. No new
functionality has been added.

## Commit 5

Introduce covid safe paths config …
3c90fba
Why:
We would like to be able to conditionally render react components based
off of the tracing strategy being used by the app. Given that the app
can be built with multiple ios targets we need to fetch this information
from over the bridge.

This commit:
Introduces the COVIDSafePathsConfig to the app, which uses the
NativeModules api to fetch the tracing strategy and stores it in a
config object to be used as needed in components.

Note that this commit only introduces the pattern to a handful of
existing components. It does not exhaustively update every spot that
needs to render conditionally based off of the tracing strategy.

At first we considered using a context for this, we rolled that back
since it seemed to be a bit overkill at this point. It might make sense
to reintroduce in a future commit if there is significant application
logic that needs to be run and shared amongst many components related to
this.

## Commit 5

Update readme, package.json for bte/gps targets …
5ccfb95
This commit updates the readme and package.json to include the info for
building to either the gps or bte target for ios


## How to test:

Run the apps locally with the two schemes.
```
yarn run-ios-gps
```

```
yarn run-ios-bte
```


Remaining work to be done:

- do a similar set of work for the android side

- verify that this setup still works with the fastlane setup and if not adjusting to do so. 
That is, I wasnt able to verify that the production builds are still working, for sake of time. It might makes sense it introduce build targets for development and staging for each of the flavors of the app.

- set up translations for bte / gps
This one could probably use some thought. Something like : t('label.my_text', 'bte') or t('label.my_text', 'gps') might makes sense. We potentially have a combinatoric problem of two app flavors for every language and setting up the translations in a maintainable way will be important.

- consider moving the swift location storage code out of AppDelegate and into the gps group so that it is only included in the gps target build.

johnschoeman committed:
Use modules imports for assets

Why:
Currently we are importing png and svg assets individually from
shared/assets. This leads to a fair amount of churn in the import
statements and is somewhat of a pain to maintain overtime. By bundling
the assets in a common module we can reduce a bit of cognitive overhead
by allowing designers and developers to always be importing assets from
same place, leading to some significant speed gains during iterations on
design implementations. Additionally,this will also allow us a place to
put some type safety if we ever introduce typescript to the project.
There is no technical gains to be realized with this strategy, it is
purely to make life a bit easier for designers.

This commit:
Introduces an /shared/assets/index.js to import icons and images from.
No new functionality has been added.

johnschoeman committed:
Refactor onboarding5 to use functional component

Why:
We would like to introduce some information to the component via a
context. It is best practice to use hooks and so we would like to use
the useContext hook for this. Currently the onboarding5 component is
build with the class syntax keeping us from doing this.

This commit:
Refactors the Onbarding5 component to use the functional style. No new
functionality has been added.

johnschoeman committed:
Rollback js code spliting

Why:
Given that we are using the two build targets of GPS and BTE on the
native side, it is no longer necessary to split the code on the
javascript side.

This commit:
Moves all of the application logic from gps/app back to app/
Additionally assets where moved out of the shared folder into app/ so
that we could remove the shared/ directory altogether. This commit is a
refactor and introduces no new functionality.

johnschoeman committed:
Update readme, package.json for bte/gps targets

This commit updates the readme and package.json to include the info for
building to either the gps or bte target for ios

johnschoeman committed:
Introduce gps and bte ios targets

Why:
We would like to introduce bluetooth to the application for bluetooth
based contact tracing. Given the privacy concerns of having both
location and bluetooth data available to the same application, google
and apple do not allow for both location and bluetooth to be available
at the same time.

Hence in order to service both location and bluetooth
we will need to provide separate apps with separate info.plists and
bundle identifiers. At the same time we would like to share a much
application logic between the two applications as to reduce the
maintenance and organizational overhead. Ideally as close to 100% of the
javascript layer is common and for the ios layer we only need to change
the info.plists.

This commit:

Introduces two build targets, GPS and BTE, to the xcode project which
are setup to read from the corresponding groups for config and info.plist
A COVIDSafePathsConfig file was introduced to expose a
'getTracingStrategy' function to the react native side so that the js
runtime can determine which tracing strategy the bundle will be using.
This will allow us flexibility to programmatically determine which
tracing strategy is being used either within a given react component or
to conditional render components altogether. This also allows use to
maximize the amount of shared native code.

To build the corresponding targets via the react native cli, you can
pass the --scheme flag with the corresponding target:

```
react-native run-ios --scheme "GPS"
react-native run-ios --scheme "BTE"
```

We updated the bundle ids to `org.pathcheck.covid-safepaths.gps` and
`org.pathcheck.covid-safepaths.bte` respectively. Depending on app store
considerations, it might be easier to revert
`org.pathcheck.covid-safepaths.gps` to `org.pathcheck.covid-safepaths`

Note that theres a fair amount of location information storing logic
currently in AppDelegate, a future commit might want to refactor this so
that the gps specific code is only built in the GPS target builds.

johnschoeman committed:
Introduce covid safe paths config

Why:
We would like to be able to conditionally render react components based
off of the tracing strategy being used by the app. Given that the app
can be built with multiple ios targets we need to fetch this information
from over the bridge.

This commit:
Introduces the COVIDSafePathsConfig to the app, which uses the
NativeModules api to fetch the tracing strategy and stores it in a
config object to be used as needed in components.

Note that this commit only introduces the pattern to a handful of
existing components. It does not exhaustively update every spot that
needs to render conditionally based off of the tracing strategy.

At first we considered using a context for this, we rolled that back
since it seemed to be a bit overkill at this point. It might make sense
to reintroduce in a future commit if there is significant application
logic that needs to be run and shared amongst many components related to
this.

Commit:
Request permissions per build flavor on android.

Commit:
Implement getTracingStrategy in Android native module. (#854)

* Introduce gps and bte ios targets

Why:
We would like to introduce bluetooth to the application for bluetooth
based contact tracing. Given the privacy concerns of having both
location and bluetooth data available to the same application, google
and apple do not allow for both location and bluetooth to be available
at the same time.

Hence in order to service both location and bluetooth
we will need to provide separate apps with separate info.plists and
bundle identifiers. At the same time we would like to share a much
application logic between the two applications as to reduce the
maintenance and organizational overhead. Ideally as close to 100% of the
javascript layer is common and for the ios layer we only need to change
the info.plists.

This commit:

Introduces two build targets, GPS and BTE, to the xcode project which
are setup to read from the corresponding groups for config and info.plist
A COVIDSafePathsConfig file was introduced to expose a
'getTracingStrategy' function to the react native side so that the js
runtime can determine which tracing strategy the bundle will be using.
This will allow us flexibility to programmatically determine which
tracing strategy is being used either within a given react component or
to conditional render components altogether. This also allows use to
maximize the amount of shared native code.

To build the corresponding targets via the react native cli, you can
pass the --scheme flag with the corresponding target:

```
react-native run-ios --scheme "GPS"
react-native run-ios --scheme "BTE"
```

We updated the bundle ids to `org.pathcheck.covid-safepaths.gps` and
`org.pathcheck.covid-safepaths.bte` respectively. Depending on app store
considerations, it might be easier to revert
`org.pathcheck.covid-safepaths.gps` to `org.pathcheck.covid-safepaths`

Note that theres a fair amount of location information storing logic
currently in AppDelegate, a future commit might want to refactor this so
that the gps specific code is only built in the GPS target builds.

* Rollback js code spliting

Why:
Given that we are using the two build targets of GPS and BTE on the
native side, it is no longer necessary to split the code on the
javascript side.

This commit:
Moves all of the application logic from gps/app back to app/
Additionally assets where moved out of the shared folder into app/ so
that we could remove the shared/ directory altogether. This commit is a
refactor and introduces no new functionality.

* Implement getTracingStrategy method inside android native module.

Co-authored-by: John Schoeman <johnschoeman1617@gmail.com>

johnschoeman committed:
Setup Fastlane for gps and bte ios targets

Why:
We would like to be able to use fastlane and match for signing and
pushing ipas for both the gps and bte targets of the app.

We would also like for to use github actions so that merges to the bte
branch will build and push the bte version of the app. This is required
to be able to ship the binaries to the app and play store, since all of
the signing certs are managed by a protected git repo that most
developers will not have access to.

Ideally we will eventually move the CD step into the same branches as
the GPS target, currently develop, release-candidate, and mater, so that
we don't have to coordinate as much between the two apps, as the vast
majority of future feature work will apply to both apps.

This commit:
Updates the fastfile and github workflow for distributing the ios
targets. We added a new set of lanes for bte_production and bte_staging.

Future work could include consolidating these lanes to build both or
either target depending on a passed parameter, though it is not clear
that this level of dryness would be worth the effort.

A more important bit of work that should probably be done in a future
commit is to automate the selection of the correct app_identifier on the
match side, currently the develop will need to comment in/out the
correct identifier.

Note that this commit does not setup any CD infrastructure for android,
this will need to happen in a future commit.

Co-Authored-By: efalkner <elizabethfalkner09@gmail.com>

Commit:
chore(release): 1.0.12

Commit:
chore(release): 1.0.13

johnschoeman committed:
Add deploy lane back to fastlane

A previous commit accidentally deleted the ios deploy lane for fastlane.
This commit adds the deploy lane back.

Commit:
chore(release): 1.0.14

tparesi committed:
Feature/bte tp onboard (#863)

* Update the onboard flow for Bluetooth

Why:
When using the BTE tracing strategy, the onboarding flow needs to change in order to
ask for bluetooth permissions. At this time, the bluetooth permissions are not properly
wired up. In the future, this will be fixed or the Exposure Notification logic from
Google/Apple will handle it.

This commit:
Adds stand-in copy for the time being to the bluetooth onboarding flow. Add bluetooth
permissions to ios. Add bluetooth logic to Onboard5.js.

Fix for info.plist relative paths issue.
Fix package.json run-ios aliases - updating the scheme

Commit:
chore(release): 1.0.15

johnschoeman committed:
Add push notification to entitlements

Why:
We are requesting to send push notifications in the info plist, but have
not enabled ASP in the app signing, which will cause the app to be
rejected during the submission process.

This commit:
Adds the push notification capability to the BTE target. We also removed
the motion tracking permission for info.plist.

Commit:
chore(release): 1.0.16

tstirrat reviewed (changes requested):
There is no way I'm going to be able to track all these changes while minimizing breakage risk..

It looks like you have a chain of things that you want to do in multiple steps, consider using [git-multi-pr](https://github.com/pingpongboss/git-multi-pr) for this, so that you can sync with `develop` and maintain multiple small prs.

It seems to match your workflow of "a single commit per logical change". 

I've used `multi-pr` once before. It assumes `master` is your base branch, but you can edit a single file to change `BASE=master` to `BASE=develop` and everything else works as intended.

johnschoeman commented:
> There is no way I'm going to be able to track all these changes while minimizing breakage risk..
> 
> It looks like you have a chain of things that you want to do in multiple steps, consider using [git-multi-pr](https://github.com/pingpongboss/git-multi-pr) for this, so that you can sync with `develop` and maintain multiple small prs.
> 
> It seems to match your workflow of "a single commit per logical change".
> 
> I've used `multi-pr` once before. It assumes `master` is your base branch, but you can edit a single file to change `BASE=master` to `BASE=develop` and everything else works as intended.

Yeah theres a lot here, and it has started to grow so that we can hit our deadlines for the bte entitlement submission.

I think it makes sense to break up into multiple prs and introduce incrementally, not sure if we need to use multi-pr, or if we should just open new prs as they make sense like with: https://github.com/Path-Check/covid-safe-paths/pull/864

So potentially this pr will become unnecessary, but I'll leave it open for now until the dust is settled for todays deadline.

jeffstolz committed:
replace ios app icons (#866)

Commit:
chore(release): 1.0.17

johnschoeman committed:
Remote links for location export and import

Why:
We would like to hide the links for the location sharing import and
export for the bte version of the app.

This commit:
Conditionally renders the links to the location import and export
screens.

Commit:
chore(release): 1.0.18

Commit:
Update fastlane (android) for bte (#868)

* Update android workflows for bte.

* Reinstate upload step.

* Run android bte publish workflow from bte branch.

* Fastfile adjustments

* Address PR feedback.

Commit:
chore(release): 1.0.19

johnschoeman committed:
Revert adding pn capacity to the BTE target.

Why:
we added the pn capacity to the bte target, since we will need push
notifications for future features. However the current provisioning
profile for the ios app does not include the ASP capacity currently. In
the meantime before we can get an updated provisioning profile, we would
like to revert the entitlement for push notifications.

This commit:
Removes the Push Notification Capacity from the ios config.

Commit:
chore(release): 1.0.20

tparesi committed:
Hide Choose Health Authorities on Settings Page (#870)

Why:
The current logic for Choosing a Health Authority involves using location.
Until the new flow is worked out, we will simply hide it.

Commit:
chore(release): 1.0.21

tparesi committed:
Differentiate GPS and BTE copy (#869)

Why:
In trying to keep the application as similar as possible, most changes are in the
onboarding flow and with copy. This PR adds logic in all places where copy differs
between the two apps. This is a good short term solution, but in the future, the
hope is to come up with a more elegant way to uncouple this logic.

This commit:

The current pattern adds a suffix of either bluetooth or location to all copy where
the two apps will differ (only in en.json).

Update COVIDSafePathsConfig to a name export.
Update Bluetooth in broadcasting service, which may be dead code at this point.

Commit:
chore(release): 1.0.22

Commit:
chore(release): 1.0.23

Commit:
chore(release): 1.0.24

Commit:
chore(release): 1.0.25

johnschoeman commented:
Closing this pr since there are things that got in to this branch which shouldnt be merged to development without some adjustment. (And it has also gotten much larger over the past few days.) Well open these changes up more incrementally, but continue working on bte as makes sense, if that sounds good to you, @mattThousand , @tparesi
